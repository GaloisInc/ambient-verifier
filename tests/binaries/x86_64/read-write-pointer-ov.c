#include <unistd.h>

/*
 * This test checks that the crucible syntax function override mechanism is
 * performing appropriately for pointer reads and writes.  It works by
 * overriding the broken `copy` function (which causes failing goals in the
 * caller by doing nothing) with a correct copy function that produces no such
 * goals.
 */

void copy(int* src, int* dest) {
  // Do nothing
}

int main(void) {
  int* res = NULL;

  int src = 1;
  int dst = 0;
  copy(&src, &dst);

  if (dst == 1) {
    res = &dst;
  };

  // NULL pointer dereference if copy failed
  return *res;
}

void _start(void) {
  main();
}
