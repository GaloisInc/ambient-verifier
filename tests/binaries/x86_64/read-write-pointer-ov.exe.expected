successful: 0
failed: 0
overrideDir: tests/overrides/
